Zend/zend_language_parser.y

```diff
@@ -1064,8 +1067,11 @@ non_empty_member_modifiers:
 
 member_modifier:
 		T_PUBLIC				{ $$ = T_PUBLIC; }
+	|	T_PUBLIC_SET			{ $$ = T_PUBLIC_SET; }
 	|	T_PROTECTED				{ $$ = T_PROTECTED; }
 	|	T_PRIVATE				{ $$ = T_PRIVATE; }
+	|	T_PROTECTED_SET			{ $$ = T_PROTECTED_SET; }
+	|	T_PRIVATE_SET			{ $$ = T_PRIVATE_SET; }
```

Comment:
Is there a reason this is tokenized and parsed as a single token, instead of the more generic `(public|protected|private)(\((set|get)(, (set|get))*\))?`?

Reply:
I assume the difficulty lies in not making `set` a keyword. It will though, break code like `const set = [1, 2, 3]; $obj->public(set);`. Which is sort of awkward that it won't work.

Reply:
Well, if that's desired then I'd rather allow it in the token itself.

Reply:
Could you at least cite the reference of the old PR related to this RFC so that people can get the history of the discussion from the beginning until now (because this RFC is very new). Isn't it too early to put it up for voting now?

This syntax is very strange to me, very unnatural.

I still think the "Hook-embedded style" is better, defining the visibility in the Hook.

The vote that defined this syntax should be restarted, because before there was no Property Hooks as a reference.

Reply:
> Well, if that's desired then I'd rather allow it in the token itself.

That would basically be the `yield from` case again. While I would never write it like that, including whitespace and comments make sense to be for syntax consistency with other places where parentheses are used. Consider the following case:

```
  public
  private(
    set, // for X
    &get, // for Y
  )
  string $foo;
```

By making it separate tokens, the syntax could also be understood in a more generic fashion by code analysis tools, e.g. formatters that want to add whitespace around parentheses.

Reply:
We already have tokens like that, e.g. `( string )`. I'm not sure it's worth complicating the parser for a theoretical future addition. I don't see us ever adding another visibility, as I mentioned in the whole syntax discussion. Other theoretical hooks (e.g. `unset` and `setRef`) should follow the `set` visibility.

Reply:
This seems like a pretty weird choice to me, especially after the intense discussion on `yield from` tokenization recently -- the impression I got from that discussion is that it would have been better to tokenize it as two tokens in the first place, so it seems weird to not follow through on that for brand-new syntax at least.

There is a technical reason why casts are tokenized as a single token, but I don't see any obvious ambiguities that would require doing so here.

It's also worth noting that this breaks the following code:
```php
<?php

class Test {
    public static function public($arg) {
        var_dump($arg);
    }
}

const set = 42;
Test::public(set);
```

Reply:
> There is a technical reason why casts are tokenized as a single token, but I don't see any obvious ambiguities that would require doing so here.

Hm, I was wrong about that. We do have an LR(1) ambiguity due to DNF types, between something like `public(set) $x` and `public (set&a)|b $x`.

Reply:
> brand-new syntax

This is just one of the problems, imagine in other parsers... Imagine when an AI analyzes the code to build the neural network, what does this neural network look like with this ambiguity in syntax?...

Since initially we will only have visibility definitions only for `get` and `set` operations, there should be no `(set)` in the second visibility definition. Implicitly, if a second definition of visibility exists, it will always be for the set operation only.

```php
class Foo
{
    public private string $bar = 'baz';
}
```

Then, later, it would be cleaner to release a new RFC to Asymmetric visibility definitions on property hooks as well, especially since new hooks may exist in the future.

Please rethink this syntax. It is not because it won a vote (which is old and occurred outside the time of this RFC) that it should be used, nor that it is the best option.
This RFC had several votes against. The author and co-author should not count because he will always be in favor.

Reply:
> Imagine when an AI analyzes the code to build the neural network, what does this neural network look like with this ambiguity in syntax?...

Parsing is a solved problem, you don't need an AI/LLM to do this.

> Please rethink this syntax.

Too late, we're past feature freeze.

> This RFC had several votes against. The author and co-author should not count because he will always be in favor.

This wouldn't have mattered. There are 24 votes in favor, 7 against. If we remove the 2 in favor from the authors, then we get 22 in favor, 7 against. This means that 22/(22+7)=75.862069% was in favor, which is above 2/3rd majority.

Reply:
> Parsing is a solved problem, you don't need an AI/LLM to do this.

I wasn't very clear. Code-generating AIs have built-in parsers and analyze code in real time as you type. Ambiguities in syntax can generate imprecise and inefficient suggestions in an autocomplete, for example.